core/src/main/java/org/elasticsearch/common/lucene/search/function/QueryFunction.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.elasticsearch.common.lucene.search.function;
+
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.search.Explanation;
+import org.apache.lucene.search.IndexSearcher;
+import org.apache.lucene.search.Query;
+import org.apache.lucene.search.Scorer;
+import org.apache.lucene.search.Weight;
+import org.apache.lucene.util.Bits;
+import org.elasticsearch.common.lucene.Lucene;
+import org.elasticsearch.index.query.functionscore.QueryFunctionBuilder;
+
+import java.io.IOException;
+import java.util.Locale;
+import java.util.Objects;
+
+/**
+ * Function that scores according to the query defined for it.
+ */
+public class QueryFunction extends ScoreFunction {
+    Query query;
+    Weight weight;
+    boolean needScores;
+
+    public QueryFunction(Query query) {
+        super(CombineFunction.MULTIPLY);
+        this.query = query;
+    }
+
+    @Override
+    public void initWeight(IndexSearcher searcher, boolean needsScores) throws IOException {
```

Comment:
maybe we should remove this method entirely and provide the IndexSearcher and create the weight in the constructor?